snap/integrity/integrity_test.go

```diff
@@ -0,0 +1,111 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package integrity_test
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/snap/integrity"
+	"github.com/snapcore/snapd/snap/squashfs"
+	"github.com/snapcore/snapd/testutil"
+)
+
+func Test(t *testing.T) { TestingT(t) }
+
+type IntegrityTestSuite struct {
+	testutil.BaseTest
+}
+
+var _ = Suite(&IntegrityTestSuite{})
+
+func (s *IntegrityTestSuite) SetUpTest(c *C) {
+	s.BaseTest.SetUpTest(c)
+}
+
+func (s *IntegrityTestSuite) TearDownTest(c *C) {
+	s.BaseTest.TearDownTest(c)
+}
+
+func (s *IntegrityTestSuite) TestAlign(c *C) {
+	align := integrity.Align
+	blockSize := uint64(integrity.BlockSize)
+
+	ret := align(0)
+	c.Check(ret, Equals, uint64(0))
+
+	ret = align(1)
+	c.Check(ret, Equals, blockSize)
+
+	ret = align(blockSize)
+	c.Check(ret, Equals, blockSize)
+
+	ret = align(uint64(blockSize + 1))
+	c.Check(ret, Equals, 2*blockSize)
+}
+
+func (s *IntegrityTestSuite) TestGenerateAndAppendSuccess(c *C) {
+	blockSize := uint64(integrity.BlockSize)
+	magic := integrity.Magic
+	snapFileName := "foo.snap"
+
+	buildDir := c.MkDir()
+	err := ioutil.WriteFile(filepath.Join(buildDir, "data.bin"), []byte("data"), 0644)
+	c.Assert(err, IsNil)
+
+	snapPath := filepath.Join(buildDir, snapFileName)
+	snap := squashfs.New(snapPath)
+	err = snap.Build(buildDir, &squashfs.BuildOpts{SnapType: "app"})
+	c.Assert(err, IsNil)
+
+	snapFileInfo, err := os.Stat(snapPath)
+	orig_size := snapFileInfo.Size()
+
+	err = integrity.GenerateAndAppend(snapPath)
+	c.Assert(err, IsNil)
+
+	snapFile, err := os.Open(snapPath)
+	c.Assert(err, IsNil)
+	defer snapFile.Close()
+
+	_, err = snapFile.Seek(orig_size, io.SeekStart)
+	c.Assert(err, IsNil)
+
+	// check integrity header
+	header := make([]byte, blockSize-1)
+	n, err := snapFile.Read(header)
+	c.Assert(err, IsNil)
+	c.Assert(n, Equals, int(blockSize)-1)
+
+	magicRead := header[0:len(magic)]
+	c.Check(magicRead, DeepEquals, magic)
+
+	metadata := header[len(magic):bytes.IndexByte(header, 0)]
+	integrityHeader := integrity.IntegrityHeader{}
+	err = json.Unmarshal(metadata, &integrityHeader)
+	c.Check(err, IsNil)
```

Comment:
Do we want a extra c.Check() here that looks at the unmashalled integretyHeader and valides some of it's properties here?

Reply:
Maybe, yes. I thought to keep it simple here and test only that we unmarshal successfully (and do more validation tests in the superblock parser once it's done), but since this function uses its own struct `IntegrityHeader` (that encapsulates the superblock) an additional test can test its extra fields. It could also guard against a change for whatever reason to any of the struct's fields that someone might put in this function in the future.